Add Ice Tea SPI communication test.

mprse opened the pull request:
### Description

Provide SPI communication test which uses IceTea Framework.

This PR can not be merged until PR which defines the new HAL API and adds at least one example implementation is merged. Probably then the test will have to be adapted to the final API version.

Detailed description can be found in README.md file in the test case folder.

**List of PRs with related tests**

| PR | Description |
| --- | --- |
| RFC SPI: Reference Implementation. #8445 | New HAL API definition and example implementations for K66F anf NUCLEO_F429ZI |
| Add SPI test and test header file. #7976 | SPI Green Tea basic test |
| Add GreeTea SPI communication test. #8216 | Two boards Green Tea SPI communication test (closed) |
| Add Ice Tea SPI communication test. #8443 | Two boards  Ice Tea SPI communication test |
| Add one board Green Tea SPI Communication Test #8919  | One board Green Tea SPI communication test |

Please assign @ithinuel as a reviewer.

### Pull request type

    [] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [X] Test update
    [ ] Breaking change

cmonr commented:
@ARMmbed/mbed-os-hal When y'all have a chance.

mprse commented:
I pushed an update of the test and documentation.
Example test output on K66F can be found below:
![example](https://user-images.githubusercontent.com/30721012/47559246-4af65b80-d915-11e8-8196-c1705dbc19c3.jpg)

Note that we are still solving test/driver issues.

mprse commented:
Looks like its ready. Please review.

Last results can be found below:
![image](https://user-images.githubusercontent.com/30721012/47718962-0be44500-dc4a-11e8-8400-ad2bcdf25c95.png)

Note that reference SPI implementation for K66F does not support lsb ordering, half duplex mode, and slave has some limitations regarding clock polarity and phase. These test cases are failing.

More details about the test can be found here:
https://github.com/mprse/mbed-os/tree/spi_feature_branch_test_com_icetea/TEST_APPS/testcases/test_spi_com

donatieng reviewed (approved):
Great work @mprse! I'll let @ithinuel confirm that he's happy with this.

cmonr commented:
/morph build

jamesbeyond commented:
@ARMmbed/mbed-os-maintainers could you put do_not_merge tag on this PR for now,
We need to consolidate this PR with Partner Enablement Team before it can be merged.
nevertheless, we do want CI run on this one to identify any potential issues.

mprse commented:
I chcecked the build failures. Symbols from the configuration are undefined, new hal api definition is missing and array sizes are not constant. Can we use only arrays with constant sizes?

mprse commented:
Modified Half-Duplex testing and added definition of port which is used for half-duplex mode to the `mbed_app.json`

mprse commented:
After discussion with PE team we agreed that this test will be shown to partners. @0xc0170 In this case please remove `DO NOT MERGE` flag. 
@jamesbeyond please confirm?

jamesbeyond commented:
Yes, as agreed during last week meeting with PE team, this can be merged

jamesbeyond commented on TEST_APPS/testcases/test_spi_com/README.md:
```diff
@@ -0,0 +1,71 @@
+
+**Test scenario**  
+
+ Test environment is shown on the picture below:  
```
Could you change diagram to proper format rather than a screenshot from PPT

jamesbeyond commented on TEST_APPS/testcases/test_spi_com/README.md:
```diff
@@ -0,0 +1,71 @@
+This test transfers data between SPI master and SPI slave.
+Two boards are required with SPI peripheral available and slave/master mode support.    
+
+To run the test using same boards as master and slave perform the following steps:  
```
A test setup section, is required here.
This should talk about the pre-requirements of the test. 
e.g. required 2 boards can be different, must support slave mode.
A simple wiring diagram needed here.
information about grand wire connection.
Any other tips about hardware preparation

mprse replied:
Fixed as suggested.

jamesbeyond commented on TEST_APPS/testcases/test_spi_com/README.md:
```diff
@@ -0,0 +1,71 @@
+ - `SPI_MASTER/SLAVE_TRANSMISSION_START_BTN`: it is possible to define button to trigger the transmission manually. In such case set `SPI_MASTER_DELAY` to 0 and define `SPI_MASTER/SLAVE_TRANSMISSION_START_BTN`. If not used then set to `NC`.  
+ - `SPI_MASTER/SLAVE_TRANSMISSION_START_LED`: set led which indicates delay before the transmission. Set `NC` if not used. 
+
+**Which SPI features are verified by the test?**  
```
which test case cover which area needs to be better explained.
Maybe this section could be moved to after the test scenario diagram. and be put into a table?
you can use[ netsocket tests readme](https://github.com/ARMmbed/mbed-os/tree/master/TESTS/netsocket) as an example.
If `Preconditions` `test steps` and `Expected results` of each test case need explanation, those can be added

mprse replied:
Fixed as suggested.

jamesbeyond reviewed (changes requested):
please see comments

jamesbeyond commented:
If the ground connection could be added to the diagram, and give one line explanation why this is required/recommended , that would be great

mprse commented:
> If the ground connection could be added to the diagram, and give one line explanation why this is required/recommended , that would be great

Sorry for the delay. I added the GND connection to the schema and the note. Also added test cases to verify buffer handling in async mode (rx len != lentx, rx or tx buf undefined) and test cases description in README file.

jamesbeyond reviewed (approved):
looks good! well done !

0xc0170 commented:
I noticed old travis script here (based on astyle changes). Can you rebase spi branch?

mprse commented:
> I noticed old travis script here (based on astyle changes). Can you rebase spi branch?

I think we need also rebase ARMmbed:feature-hal-spec-spi since there is currently the issue with building Mbed on Windows. 
I prepared this branch few days ago (feature-hal-spec-spi rebased on master without the issue):
https://github.com/mprse/mbed-os/commits/feature-hal-spec-spi

Also this test requires this PRs:
PR https://github.com/ARMmbed/mbed-os/pull/8445
PR https://github.com/ithinuel/mbed-os/pull/5 (some updates to #8445)
We should move forward with this one. It would be much easier to maintain the tests.

mprse commented:
> Can you rebase spi branch?

Rebased.

0xc0170 commented:
CI started

jupe reviewed (commented):
Just curious, is there some reason why master and slave could not be compiled to same app? Would it increase flash size too much or? Combining both modes would safe some CI time, simplify test configuration a bit and reduce maintenance work as well because less code to keep up-to-date..

cmonr commented:
@mprse Please take a look at the following IAR build failures:
http://mbed-os-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/builds/8443/6/build-IAR/

mprse commented:
> Just curious, is there some reason why master and slave could not be compiled to same app? Would it increase flash size too much or? Combining both modes would safe some CI time, simplify test configuration a bit and reduce maintenance work as well because less code to keep up-to-date..

Thanks for the tip. There is no specific reason for that. It should be easy to combine both modes. I'll check that.

mprse committed:
Add Ice Tea SPI communication test.

mprse commented:
> Just curious, is there some reason why master and slave could not be compiled to same app? Would it increase flash size too much or? Combining both modes would safe some CI time, simplify test configuration a bit and reduce maintenance work as well because less code to keep up-to-date..

Updated the test as suggested. Now same app is used for SPI master board and SPI slave board. Updated the documentation also.

mprse commented:
> @mprse Please take a look at the following IAR build failures:
http://mbed-os-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/builds/8443/6/build-IAR/

Checked the build failures. Most errors are because new HAL API (PR #8445) is not on the feature branch yet:
```
        [Error] spi_test_common.h@88,6: [Pe020]: identifier "spi_mode_t" is undefined
        [Error] spi_test_common.h@89,14: [Pe020]: identifier "spi_bit_ordering_t" is undefined
```

Fixed other problems.

mprse commented:
Current test output for NUCLEO_F429ZI (half duplex test case requires different wire connection):
![image](https://user-images.githubusercontent.com/30721012/49376663-412bf900-f708-11e8-8c45-cf35d059b3e3.png)

0xc0170 commented:
@mprse Looks like this one is ready for CI?

cmonr commented:
@donatieng @ithinuel @jamesbeyond All good? Looks like CI is ready to be started.

cmonr commented:
@mprse When you get a chance, please rebase your PR. 

I updated the base branch with fixes for Travis CI.

mprse commented:
This PR requres PR https://github.com/ARMmbed/mbed-os/pull/8445

cmonr commented:
@MarceloSalazar @screamerbg Could y'all look this over sometime soon?

@mprse This needs a rebase.

cmonr commented:
@mprse Rebase when available.

mprse commented:
Rebased after reference impl has been merged into the feature branch.
This one is ready for review and CI.

cmonr commented:
@mprse Please rebase again (sorry). The base branch has been rebased and updated.

mprse commented:
> @mprse Please rebase again (sorry). The base branch has been rebased and updated.

Yes, I asked @0xc0170 to rebase spi feature branch because of the Travis issues.

mprse commented:
Rebased after spi feature branch has been rebased on master to solve Travis issue.

mprse commented:
Rebased and fixed style.
